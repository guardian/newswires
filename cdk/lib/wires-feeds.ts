import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuCname } from '@guardian/cdk/lib/constructs/dns';
import {
	GuParameter,
	GuStack,
	GuStringParameter,
} from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import { GuS3Bucket } from '@guardian/cdk/lib/constructs/s3';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { InstanceClass, InstanceSize, InstanceType } from 'aws-cdk-lib/aws-ec2';
import { ArnPrincipal, User } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { SqsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { SqsSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import { GuDatabase } from './constructs/database';
import { Alarms, GuPlayApp } from '@guardian/cdk';
import { NoMonitoring } from '@guardian/cdk/lib/constructs/cloudwatch';
import { AccessScope } from '@guardian/cdk/lib/constants';
import { GuGetS3ObjectsPolicy } from '@guardian/cdk/lib/constructs/iam';
import { GuVpc, SubnetType } from '@guardian/cdk/lib/constructs/ec2';

export type WiresFeedsProps = GuStackProps & {
	domainName: string;
	enableMonitoring: boolean;
};
const app = 'wires-feeds';

export class WiresFeeds extends GuStack {
	constructor(scope: App, id: string, props: WiresFeedsProps) {
		super(scope, id, { ...props, app });

		const { domainName, enableMonitoring } = props;

		const privateSubnets = GuVpc.subnetsFromParameter(this, {
			type: SubnetType.PRIVATE,
			app,
		});

		const stageStackApp = `${this.stage}/${this.stack}/${app}`;

		const fingerpostPublishingUserArn = new GuStringParameter(
			this,
			'fingerpost-publishing-user-arn',
			{
				fromSSM: true,
				default: `/${stageStackApp}/fingerpost-publishing-user-arn`,
			},
		).valueAsString;

		const user = User.fromUserArn(
			this,
			'fingerpost-publishing-user',
			fingerpostPublishingUserArn,
		);

		function createTopicQueue(scope: GuStack, topicType: string) {
			const topic = new Topic(scope, `${topicType}-topic`, {
				enforceSSL: true,
			});
			topic.grantPublish(new ArnPrincipal(user.userArn));

			const queueName = `${topicType}-queue`;
			const visibilityTimeout = Duration.minutes(5);

			const deadLetterQueue = new Queue(
				scope,
				`${queueName}DeadLetterQueue-${props.stage}`,
				{ visibilityTimeout },
			);

			const queue = new Queue(scope, queueName, {
				enforceSSL: true,
				retentionPeriod: Duration.days(14),
				// We are using this queue in conjunction with a lambda SqsEventSource
				// visibilityTimeout is set by default to 5 minutes to ensure that the lambda has
				// enough time to process the message before it becomes visible again.
				visibilityTimeout: Duration.minutes(5),
				deadLetterQueue: {
					queue: deadLetterQueue,
					maxReceiveCount: 3,
				},
			});

			topic.addSubscription(
				new SqsSubscription(queue, { rawMessageDelivery: true }),
			);

			return queue;
		}

		const databaseName = 'newswires';

		const database = new GuDatabase(this, 'NewswiresDB', {
			app,
			instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.SMALL),
			allowExternalConnection: true,
			databaseName,
			multiAz: false,
			devxBackups: true,
			vpcSubnets: {
				subnets: privateSubnets,
			},
		});

		/** A topic and queue for the 'raw' wires feed.
		 * Not receiving data yet so we aren't currently doing anything more with it.
		 */
		createTopicQueue(this, 'source');

		const fingerPostQueue = createTopicQueue(this, 'fingerpost');

		const feedsBucket = new GuS3Bucket(this, `feeds-bucket-${this.stage}`, {
			app,
			versioned: true,
		});

		const ingestionLambda = new GuLambdaFunction(
			this,
			`IngestionLambda-${this.stage}`,
			{
				app: 'ingestion-lambda',
				runtime: Runtime.NODEJS_20_X,
				handler: 'handler.main',
				fileName: 'ingestion-lambda.zip',
				environment: {
					FEEDS_BUCKET_NAME: feedsBucket.bucketName,
					DATABASE_ENDPOINT_ADDRESS: database.dbInstanceEndpointAddress,
					DATABASE_PORT: database.dbInstanceEndpointPort,
					DATABASE_NAME: databaseName,
				},
			},
		);

		const eventSource = new SqsEventSource(fingerPostQueue, {
			/**
			 * This is required to allow us to deal with failures of particular
			 * records handed to a lambda by an SQS queue. Without this, the
			 * lambda will retry the entire batch of records, which is not what
			 * we want.
			 *
			 * See https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html#services-sqs-batchfailurereporting
			 *
			 */
			reportBatchItemFailures: true,
		});

		ingestionLambda.addEventSource(eventSource);

		feedsBucket.grantWrite(ingestionLambda);

		database.grantConnect(ingestionLambda);

		const panDomainSettingsBucket = new GuParameter(
			this,
			'PanDomainSettingsBucket',
			{
				description: 'Bucket name for pan-domain auth settings',
				fromSSM: true,
				default: `/${stageStackApp}/pan-domain-settings-bucket`,
				type: 'String',
			},
		);

		const permissionsBucketName = new GuParameter(this, 'PermissionsBucket', {
			description: 'Bucket name for permissions data',
			fromSSM: true,
			default: `/${stageStackApp}/permissions-bucket`,
			type: 'String',
		});

		const alarmSnsTopic = new Topic(this, `${app}-email-alarm-topic`);

		const scaling = {
			minimumInstances: 1,
			maximumInstances: 2,
		};

		const monitoringConfiguration: Alarms | NoMonitoring = enableMonitoring
			? {
					snsTopicName: alarmSnsTopic.topicName,
					unhealthyInstancesAlarm: true,
					http5xxAlarm: {
						tolerated5xxPercentage: 10,
						numberOfMinutesAboveThresholdBeforeAlarm: 1,
					},
				}
			: { noMonitoring: true };

		const nwApp = 'newswires';

		const newswiresApp = new GuPlayApp(this, {
			app,
			access: { scope: AccessScope.PUBLIC },
			privateSubnets,
			instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.SMALL),
			certificateProps: {
				domainName,
			},
			monitoringConfiguration,
			userData: {
				distributable: {
					fileName: `${nwApp}.deb`,
					executionStatement: `dpkg -i /${app}/${nwApp}.deb`, // TODO fix pls
				},
			},
			scaling,
			applicationLogging: { enabled: true, systemdUnitName: nwApp },
			imageRecipe: 'editorial-tools-focal-java17-ARM-WITH-cdk-base',
			roleConfiguration: {
				additionalPolicies: [
					new GuGetS3ObjectsPolicy(this, 'PandaAuthPolicy', {
						bucketName: panDomainSettingsBucket.valueAsString,
					}),
					new GuGetS3ObjectsPolicy(this, 'PermissionsCachePolicy', {
						bucketName: permissionsBucketName.valueAsString,
						paths: [`${this.stage}/*`],
					}),
				],
			},
		});

		// Add the domain name
		new GuCname(this, 'DnsRecord', {
			app: nwApp,
			domainName: domainName,
			ttl: Duration.minutes(1),
			resourceRecord: newswiresApp.loadBalancer.loadBalancerDnsName,
		});

		database.grantConnect(newswiresApp.autoScalingGroup);
		newswiresApp.autoScalingGroup.connections.addSecurityGroup(
			database.accessSecurityGroup,
		);
	}
}
