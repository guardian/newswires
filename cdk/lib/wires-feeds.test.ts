import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { WiresFeeds } from './wires-feeds';

describe('The WiresFeeds stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new WiresFeeds(app, 'WiresFeeds', {
			stack: 'editorial-feeds',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
