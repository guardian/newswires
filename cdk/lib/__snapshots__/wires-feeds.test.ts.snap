// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The WiresFeeds stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStringParameter",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "fingerpostpublishinguserarn": {
      "Default": "/TEST/editorial-feeds/wires-feeds/fingerpost-publishing-user-arn",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "fingerpostqueueDeadLetterQueueTESTC46C734A": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "App",
            "Value": "wires-feeds",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newswires",
          },
          {
            "Key": "Stack",
            "Value": "editorial-feeds",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "fingerpostqueueF09A4E0A": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "fingerpostqueueDeadLetterQueueTESTC46C734A",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "wires-feeds",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newswires",
          },
          {
            "Key": "Stack",
            "Value": "editorial-feeds",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Retain",
    },
    "fingerpostqueuePolicy8A8FE8AE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "fingerpostqueueF09A4E0A",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "fingerposttopic0A8134FC",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "fingerpostqueueF09A4E0A",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "fingerpostqueueF09A4E0A",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "fingerpostqueueWiresFeedsfingerposttopic09B5AE1CD1B1A0A5": {
      "DependsOn": [
        "fingerpostqueuePolicy8A8FE8AE",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "fingerpostqueueF09A4E0A",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": true,
        "TopicArn": {
          "Ref": "fingerposttopic0A8134FC",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "fingerposttopic0A8134FC": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "Tags": [
          {
            "Key": "App",
            "Value": "wires-feeds",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newswires",
          },
          {
            "Key": "Stack",
            "Value": "editorial-feeds",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
      "UpdateReplacePolicy": "Retain",
    },
    "fingerposttopicPolicy93AB9BE9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": "*",
              "Resource": {
                "Ref": "fingerposttopic0A8134FC",
              },
              "Sid": "AllowPublishThroughSSLOnly",
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Ref": "fingerpostpublishinguserarn",
                },
              },
              "Resource": {
                "Ref": "fingerposttopic0A8134FC",
              },
              "Sid": "1",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": [
          {
            "Ref": "fingerposttopic0A8134FC",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "sourcequeue87B8A558": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "sourcequeueDeadLetterQueueTEST4E4374EA",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "wires-feeds",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newswires",
          },
          {
            "Key": "Stack",
            "Value": "editorial-feeds",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Retain",
    },
    "sourcequeueDeadLetterQueueTEST4E4374EA": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "App",
            "Value": "wires-feeds",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newswires",
          },
          {
            "Key": "Stack",
            "Value": "editorial-feeds",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "sourcequeuePolicyD21DF4F1": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "sourcequeue87B8A558",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "sourcetopic7C3DC892",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "sourcequeue87B8A558",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "sourcequeue87B8A558",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "sourcequeueWiresFeedssourcetopic2270C481B5A95A06": {
      "DependsOn": [
        "sourcequeuePolicyD21DF4F1",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "sourcequeue87B8A558",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": true,
        "TopicArn": {
          "Ref": "sourcetopic7C3DC892",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "sourcetopic7C3DC892": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "Tags": [
          {
            "Key": "App",
            "Value": "wires-feeds",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newswires",
          },
          {
            "Key": "Stack",
            "Value": "editorial-feeds",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
      "UpdateReplacePolicy": "Retain",
    },
    "sourcetopicPolicy693F1C42": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": "*",
              "Resource": {
                "Ref": "sourcetopic7C3DC892",
              },
              "Sid": "AllowPublishThroughSSLOnly",
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Ref": "fingerpostpublishinguserarn",
                },
              },
              "Resource": {
                "Ref": "sourcetopic7C3DC892",
              },
              "Sid": "1",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": [
          {
            "Ref": "sourcetopic7C3DC892",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
  },
}
`;
